File: src/main/java/net/arvandor/numinoustreasury/listener/PlayerMoveListener.java
package net.arvandor.numinoustreasury.listener;

import com.rpkit.core.service.Services;
import net.arvandor.numinoustreasury.node.NuminousNode;
import net.arvandor.numinoustreasury.node.NuminousNodeService;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;

public final class PlayerMoveListener implements Listener {

    @EventHandler
    public void onPlayerMove(PlayerMoveEvent event) {
        NuminousNodeService nodeService = Services.INSTANCE.get(NuminousNodeService.class);
        NuminousNode nodeWithEntranceAtLocation = nodeService.getNodeWithEntranceAtLocation(event.getTo());
        if (nodeWithEntranceAtLocation != null) {
            event.setTo(nodeWithEntranceAtLocation.getEntranceWarpDestination());
            return;
        }
        NuminousNode nodeWithExitAtLocation = nodeService.getNodeWithExitAtLocation(event.getTo());
        if (nodeWithExitAtLocation != null) {
            event.setTo(nodeWithExitAtLocation.getExitWarpDestination());
            return;
        }
    }

}

File: src/main/kotlin/net/arvandor/numinoustreasury/listener/PlayerMoveListener.kt
package net.arvandor.numinoustreasury.listener

import com.rpkit.core.service.Services
import net.arvandor.numinoustreasury.node.NuminousNodeService
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent

class PlayerMoveListener : Listener {
    @EventHandler
    fun onPlayerMove(event: PlayerMoveEvent) {
        val nodeService =
            Services.INSTANCE.get(
                NuminousNodeService::class.java,
            )
        val toLocation = event.to
        if (toLocation != null) {
            val nodeWithEntranceAtLocation = nodeService.getNodeWithEntranceAtLocation(toLocation)
            if (nodeWithEntranceAtLocation != null) {
                event.setTo(nodeWithEntranceAtLocation.entranceWarpDestination)
                return
            }
            val nodeWithExitAtLocation = nodeService.getNodeWithExitAtLocation(toLocation)
            if (nodeWithExitAtLocation != null) {
                event.setTo(nodeWithExitAtLocation.exitWarpDestination)
                return
            }
        }
    }
}
